presto-plugin-toolkit/src/main/java/com/facebook/presto/security/RefreshingAccessControlUtils.java

```diff
@@ -0,0 +1,102 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public final class RefreshingAccessControlUtils
+{
+    private RefreshingAccessControlUtils()
+    {
+    }
+
+    public static SystemAccessControl refreshingSystemAccessControl(Supplier<SystemAccessControl> factory, Duration refreshPeriod)
+    {
+        return new DelegateSystemAccessControl(new Refresher<>(factory, () -> true, refreshPeriod));
+    }
+
+    public static SystemAccessControl refreshingSystemAccessControl(Supplier<SystemAccessControl> factory, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+    {
+        return new DelegateSystemAccessControl(new Refresher<>(factory, refreshCondition, refreshPeriod));
+    }
+
+    public static ConnectorAccessControl refreshingConnectorAccessControl(Supplier<ConnectorAccessControl> factory, Duration refreshPeriod)
+    {
+        return new DelegateConnectorAccessControl(new Refresher<>(factory, () -> true, refreshPeriod));
+    }
+
+    public static ConnectorAccessControl refreshingConnectorAccessControl(Supplier<ConnectorAccessControl> factory, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+    {
+        return new DelegateConnectorAccessControl(new Refresher<>(factory, refreshCondition, refreshPeriod));
+    }
+
+    public static Supplier<Boolean> wasFileModified(Path path)
+    {
+        AtomicLong configFileNameLastModification = new AtomicLong();
+        return () -> {
+            long lastModification = path.toFile().lastModified();
+            return configFileNameLastModification.get() != lastModification &&
+                    configFileNameLastModification.compareAndSet(configFileNameLastModification.get(), lastModification);
+        };
+    }
+
+    private static final class Refresher<T>
+            implements Supplier<T>
+    {
+        private final Supplier<T> supplier;
+        private final Supplier<Boolean> refreshCondition;
+        private final Duration refreshPeriod;
+        private final AtomicReference<T> object = new AtomicReference<>();
+        private final AtomicLong lastRefreshNanos = new AtomicLong(System.nanoTime());
+
+        public Refresher(Supplier<T> supplier, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+        {
+            this.supplier = requireNonNull(supplier, "supplier is null");
+            this.refreshCondition = requireNonNull(refreshCondition, "refreshCondition is null");
+            this.refreshPeriod = requireNonNull(refreshPeriod, "refreshPeriod is null");
+            object.set(this.supplier.get());
+        }
+
+        @Override
+        public T get()
+        {
+            if (Duration.nanosSince(lastRefreshNanos.get()).compareTo(refreshPeriod) > 0) {
+                refresh();
+            }
+            return object.get();
+        }
+
+        private void refresh()
```

Comment:
as you observed (and i am only noting), when `refresh` fails, you end up using previous implementation indefinitely.
This can be very easily fixed by synchronizing `get` and storing last refresh time _after_ loading new impl.